29_Arithmetic2Bench/main.cpp

```diff
@@ -0,0 +1,739 @@
+			}
+
+			auto surface = CSurfaceVulkanWin32::create(smart_refctd_ptr(m_api), smart_refctd_ptr_static_cast<IWindowWin32>(m_window));
+			const_cast<std::remove_const_t<decltype(m_surface)>&>(m_surface) = CSimpleResizeSurface<ISimpleManagedSurface::ISwapchainResources>::create(std::move(surface));
+		}
+
+		if (m_surface)
+			return { {m_surface->getSurface()/*,EQF_NONE*/} };
+
+		return {};
+	}
+
+	bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+	{
+		m_inputSystem = make_smart_refctd_ptr<InputSystem>(logger_opt_smart_ptr(smart_refctd_ptr(m_logger)));
+
+		if (!device_base_t::onAppInitialized(std::move(system)))
+			return false;
+		if (!asset_base_t::onAppInitialized(std::move(system)))
+			return false;
+
+		m_semaphore = m_device->createSemaphore(m_realFrameIx);
+		if (!m_semaphore)
+			return logFail("Failed to Create a Semaphore!");
+
+		ISwapchain::SCreationParams swapchainParams = { .surface = m_surface->getSurface() };
+		if (!swapchainParams.deduceFormat(m_physicalDevice))
+			return logFail("Could not choose a Surface Format for the Swapchain!");
+
+		auto graphicsQueue = getGraphicsQueue();
+		if (!m_surface || !m_surface->init(graphicsQueue, std::make_unique<ISimpleManagedSurface::ISwapchainResources>(), swapchainParams.sharedParams))
+			return logFail("Could not create Window & Surface or initialize the Surface!");
+
+		auto pool = m_device->createCommandPool(graphicsQueue->getFamilyIndex(), IGPUCommandPool::CREATE_FLAGS::RESET_COMMAND_BUFFER_BIT);
+
+		for (auto i = 0u; i < MaxFramesInFlight; i++)
+		{
+			if (!pool)
+				return logFail("Couldn't create Command Pool!");
+			if (!pool->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY, { m_cmdBufs.data() + i, 1 }))
+				return logFail("Couldn't create Command Buffer!");
+		}
+
+		m_winMgr->setWindowSize(m_window.get(), WIN_W, WIN_H);
+		m_surface->recreateSwapchain();
+
+		transferDownQueue = getTransferDownQueue();
+		computeQueue = getComputeQueue();
+
+		// TODO: get the element count from argv
+		const uint32_t elementCount = Output<>::ScanElementCount;
+		// populate our random data buffer on the CPU and create a GPU copy
+		inputData = new uint32_t[elementCount];
+		{
+			std::mt19937 randGenerator(0xdeadbeefu);
+			for (uint32_t i = 0u; i < elementCount; i++)
+				inputData[i] = randGenerator(); // TODO: change to using xoroshiro, then we can skip having the input buffer at all
+
+			IGPUBuffer::SCreationParams inputDataBufferCreationParams = {};
+			inputDataBufferCreationParams.size = sizeof(Output<>::data[0]) * elementCount;
+			inputDataBufferCreationParams.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT | IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT;
+			m_utils->createFilledDeviceLocalBufferOnDedMem(
+				SIntendedSubmitInfo{.queue=getTransferUpQueue()},
+				std::move(inputDataBufferCreationParams),
+				inputData
+			).move_into(gpuinputDataBuffer);
+		}
```

Comment:
use pcg or other hash in the shader and get rid of the input buffer, it will reduce the data access overhead

Reply:
https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/190#discussion_r2036198721